Fix nested render without rendering block

chrisgibson opened the pull request:
I ran into the same issue as reported in https://github.com/bullet-train-co/nice_partials/issues/98. `local_assigns` aren't being set on nested renderings when the call to render doesn't supply a block.

I narrowed it down to this line: https://github.com/bullet-train-co/nice_partials/blob/fc66c50c8828bcb374f5f3abd0acb0a43c014085/lib/nice_partials/monkey_patch.rb#L113
`capture_with_outer_partial_access` isn't being called when there isn't a supplied block.

This change in this PR moves the block guard from within the `render_partial_template` method to the `capture_with_outer_partial_access` method.

chrisgibson committed:
Fix nested render without rendering block

Make use of partials stack even when there is no rendering block.

kaspth commented:
@chrisgibson thank you very much for the PR! And to @adampal for reporting it originally.

I haven't had as much time for Nice Partials for a while. I'm hoping to do a double check of this PR this week.

Then I'll merge and ship a release. Thanks again!

kaspth committed:
Replace complex stack state with a linked list

When we're rendering deeply nested partials we're essentially pushing and popping off of a stack of partials.

This worked in cases like:

```erb
<%# app/views/_outer.html.erb %>
<%# There's a `partial` here that we want to refer to within the render block: %>

<%= render "card" do |card| %>
  <%= card.title partial.title %> <%# …so we ensure that `partial` here still refers to the outer partial. %>
<% end %>

<%# Now that we're outside the block, `partial` still points to the original outer calling partial. %>
```

However, when we're rendering the innerds of `card` itself, calling `partial` there should return itself and not the `_outer` partial.

We were relying on inner partials being called with a block, which would end up calling `__partials.reset_locator` and it'd sorta work.

If however, we rendered an inner partial within the block, without its own block, we wouldn't call `reset_locator` and passing any `local_assigns` like `body:` wouldn't be recognized.
Like this:

```erb
<%= render "card" do |card| %>
  <%= render "other/partial", body: "Lorem Ipsum" %>
<% end %>
```

This whole stack implementation is quite needlessly complex, and I never liked it (I wrote it so I can say that),
and turns out it works better if we convert our implementation to just have partials know about their outer partial.

And then our rendering code becomes a little simpler too.

This also means that `card.outer_partial` will return the `_outer` partial too!

This may open the door to easier communication between partials, because we can communicate all the way up the partial stack now!

kaspth commented:
@chrisgibson thanks again! I played around with the implementation some more and felt like `capture_with_outer_partial_access` shouldn't be called without a block.

I ended up coming up with an alternative that finally called removed our whole Stack implementation, which I always felt icky about. I tried explaining more thoroughly in the commit, I hope it makes sense!

kaspth commented:
I was planning on shipping a release for this, but it looks like our CI isn't quite fresh anymore. https://github.com/bullet-train-co/nice_partials/commit/06eae3f84f9ce0d3d7a6623314eb66f26a005916

I'm not sure why or how we're getting Ruby 2.7.8 there.

kaspth commented:
Just released this, thanks again @chrisgibson and @adampal for reporting! Let me know how it goes https://github.com/bullet-train-co/nice_partials/releases/tag/v0.10.1
